Add visualization of 3D particle emission shapes

paddy-exe opened the pull request:
Adds support for visualization of particle emission shapes. 

Separated part of https://github.com/godotengine/godot/pull/86902 due to discussion with @QbieShay on RocketChat

## TODO
- [x] Implement support for ring_cone_angle (with support of @Arnklit) 

## Showcase
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/d175b065-ba15-40e1-8b60-9a2eeb70cf44" />

paddy-exe committed:
Add visualization of 3D particle emission shapes

Co-authored-by: Kasper Arnklit Frandsen <kasper.arnklit@gmail.com>

QbieShay commented:
Should the signal emission be gated behind a tools build?

paddy-exe commented:
> Should the signal emission be gated behind a tools build?

Hmm I guess we could put it behind a tools build flag 🤔 not sure if we have to though. Do you think we should?

Calinou commented:
> Hmm I guess we could put it behind a tools build flag 🤔 not sure if we have to though. Do you think we should?

I'd say yes, if the signal has no purpose outside of the editor.

QbieShay commented:
My rationale here is that I don't want this to cause performance issues in case people are animating the size (for whatever reason).

Alternatively instead of using a signal, it could be polled, but i don't know if gizmos support it.

paddy-exe commented:
@QbieShay I think I added this now

clayjohn commented on scene/resources/particle_process_material.cpp:
```diff
@@ -2209,6 +2235,8 @@ void ParticleProcessMaterial::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "sub_emitter_amount_at_collision", PROPERTY_HINT_RANGE, "1,32,1"), "set_sub_emitter_amount_at_collision", "get_sub_emitter_amount_at_collision");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "sub_emitter_keep_velocity"), "set_sub_emitter_keep_velocity", "get_sub_emitter_keep_velocity");
 
+	ADD_SIGNAL(MethodInfo("emission_shape_changed"));
```
Let's remove this so we don't have to emit the signal at runtime

paddy-exe replied:
Removed 👍🏻

Arnklit commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,304 @@
+					float rb = Math::deg_to_rad((float)((i + 1) * 3));
+					Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
+					Point2 b = Vector2(Math::sin(rb), Math::cos(rb)) * ring_radius;
+					Point2 inner_a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_inner_radius;
+					Point2 inner_b = Vector2(Math::sin(rb), Math::cos(rb)) * ring_inner_radius;
+
+					points.push_back(basis.xform(Vector3(a.x * scale.x + offset.x, half_ring_height * scale.y + offset.y, a.y * scale.z + offset.z)));
+					points.push_back(basis.xform(Vector3(b.x * scale.x + offset.x, half_ring_height * scale.y + offset.y, b.y * scale.z + offset.z)));
+					// Outer top ring cap.
+					// Outer bottom ring cap.
+					points.push_back(basis.xform(Vector3(a.x * scale.x + offset.x, -half_ring_height * scale.y + offset.y, a.y * scale.z + offset.z)));
+					points.push_back(basis.xform(Vector3(b.x * scale.x + offset.x, -half_ring_height * scale.y + offset.y, b.y * scale.z + offset.z)));
+
+					// Inner top ring cap.
+					points.push_back(basis.xform(Vector3(inner_a.x * scale.x + offset.x, half_ring_height * scale.y + offset.y, inner_a.y * scale.z + offset.z)));
+					points.push_back(basis.xform(Vector3(inner_b.x * scale.x + offset.x, half_ring_height * scale.y + offset.y, inner_b.y * scale.z + offset.z)));
+					// Inner bottom ring cap.
+					points.push_back(basis.xform(Vector3(inner_a.x * scale.x + offset.x, -half_ring_height * scale.y + offset.y, inner_a.y * scale.z + offset.z)));
+					points.push_back(basis.xform(Vector3(inner_b.x * scale.x + offset.x, -half_ring_height * scale.y + offset.y, inner_b.y * scale.z + offset.z)));
+				}
+
+				for (int i = 0; i <= 120; i = i + 30) {
+					float ra = Math::deg_to_rad((float)(i * 3));
+					Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
+					Point2 inner_a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_inner_radius;
+
+					// Outer 90 degrees vertical lines.
+					points.push_back(basis.xform(Vector3(a.x * scale.x + offset.x, half_ring_height * scale.y + offset.y, a.y * scale.z + offset.z)));
+					points.push_back(basis.xform(Vector3(a.x * scale.x + offset.x, -half_ring_height * scale.y + offset.y, a.y * scale.z + offset.z)));
+
+					// Inner 90 degrees vertical lines.
+					points.push_back(basis.xform(Vector3(inner_a.x * scale.x + offset.x, half_ring_height * scale.y + offset.y, inner_a.y * scale.z + offset.z)));
+					points.push_back(basis.xform(Vector3(inner_a.x * scale.x + offset.x, -half_ring_height * scale.y + offset.y, inner_a.y * scale.z + offset.z)));
+				}
+
+				if (p_gizmo->is_selected()) {
+					p_gizmo->add_lines(points, material);
+				}
+			}
+		}
```
Fixes for the ring gizmo
```suggestion
			} else if (shape == ParticleProcessMaterial::EMISSION_SHAPE_RING) {
				Vector3 offset = mat->get_emission_shape_offset();
				Vector3 scale = mat->get_emission_shape_scale();

				float ring_height = mat->get_emission_ring_height();
				float half_ring_height = ring_height / 2;
				float ring_radius = mat->get_emission_ring_radius();

				float ring_inner_radius = mat->get_emission_ring_inner_radius();
				Vector3 ring_axis = mat->get_emission_ring_axis();
				float ring_cone_angle = mat->get_emission_ring_cone_angle();

				// top radius stuff
				float ring_radius_top = MAX(ring_radius - Math::tan(Math::deg_to_rad(90.0 - ring_cone_angle)) * ring_height, 0.0);
				float ring_inner_radius_top = (ring_inner_radius / ring_radius) * ring_radius_top;

				Vector<Vector3> points;

				Basis basis;
				basis.rows[1] = ring_axis.normalized();
				basis.rows[0] = Vector3(basis[1][1], -basis[1][2], -basis[1][0]).normalized();
				basis.rows[0] = (basis[0] - basis[0].dot(basis[1]) * basis[1]).normalized();
				basis[2] = basis[0].cross(basis[1]).normalized();
				
				//offset = basis.xform(offset);
				//scale = basis.xform(scale);
				//basis = basis.orthonormalized();
				basis = basis.inverse();

				for (int i = 0; i <= 120; i++) {
					float ra = Math::deg_to_rad((float)(i * 3));
					float rb = Math::deg_to_rad((float)((i + 1) * 3));
					Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
					Point2 b = Vector2(Math::sin(rb), Math::cos(rb)) * ring_radius;
					Point2 a2 = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius_top;
					Point2 b2 = Vector2(Math::sin(rb), Math::cos(rb)) * ring_radius_top;
					Point2 inner_a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_inner_radius;
					Point2 inner_b = Vector2(Math::sin(rb), Math::cos(rb)) * ring_inner_radius;
					Point2 inner_a2 = Vector2(Math::sin(ra), Math::cos(ra)) * ring_inner_radius_top;
					Point2 inner_b2 = Vector2(Math::sin(rb), Math::cos(rb)) * ring_inner_radius_top;

					// Outer top ring cap.
					points.push_back(basis.xform(Vector3(a2.x, half_ring_height, a2.y)) * scale + offset);
					points.push_back(basis.xform(Vector3(b2.x, half_ring_height, b2.y)) * scale + offset);
					
					// Outer bottom ring cap.
					points.push_back(basis.xform(Vector3(a.x, -half_ring_height, a.y)) * scale + offset);
					points.push_back(basis.xform(Vector3(b.x, -half_ring_height, b.y)) * scale + offset);

					// Inner top ring cap.
					points.push_back(basis.xform(Vector3(inner_a2.x, half_ring_height, inner_a2.y)) * scale + offset);
					points.push_back(basis.xform(Vector3(inner_b2.x, half_ring_height, inner_b2.y)) * scale + offset);
					
					// Inner bottom ring cap.
					points.push_back(basis.xform(Vector3(inner_a.x, -half_ring_height, inner_a.y)) * scale + offset);
					points.push_back(basis.xform(Vector3(inner_b.x, -half_ring_height, inner_b.y)) * scale + offset);
				}

				for (int i = 0; i <= 360; i = i + 90) {
					float ra = Math::deg_to_rad((float)(i));
					Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
					Point2 a2 = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius_top;
					Point2 inner_a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_inner_radius;
					Point2 inner_a2 = Vector2(Math::sin(ra), Math::cos(ra)) * ring_inner_radius_top;

					// Outer 90 degrees vertical lines.
					points.push_back(basis.xform(Vector3(a2.x , half_ring_height, a2.y)) * scale + offset);
					points.push_back(basis.xform(Vector3(a.x, -half_ring_height, a.y)) * scale + offset);

					// Inner 90 degrees vertical lines.
					points.push_back(basis.xform(Vector3(inner_a2.x, half_ring_height, inner_a2.y)) * scale + offset);
					points.push_back(basis.xform(Vector3(inner_a.x, -half_ring_height, inner_a.y)) * scale + offset);
				}

				if (p_gizmo->is_selected()) {
					p_gizmo->add_lines(points, material);
				}
			}
		}
```

Arnklit commented:
The above suggestion should correct the ring gizmo. Needs to be added for CPU as well, but it should be identical.

https://github.com/user-attachments/assets/742103f1-81bf-4f79-9fb1-5b4f90f0e49c

paddy-exe commented:
- [x] Added commit from Kasper
- [x] Added CPU Particle support for ring cone angle
<img width="1437" alt="image" src="https://github.com/user-attachments/assets/df3c5608-8cc7-443d-b89a-a2d9d9c0875e" />

- [x] Cleaned up the code comments

Ready for review now 👍🏻

Mickeon commented on scene/resources/particle_process_material.cpp:
```diff
@@ -1567,16 +1567,25 @@ void ParticleProcessMaterial::set_emission_shape(EmissionShape p_shape) {
 	emission_shape = p_shape;
 	notify_property_list_changed();
 	_queue_shader_change();
+#ifdef TOOLS_ENABLED
+	emit_changed();
+#endif
```
All affected properties should have this documented, because emitting the signal only in the editor is an exception. 
Whether it should at project run-time as well is debatable. If anything, all other particle properties should make note that `changed` is not emitted for them, but that's for another time.

Something like `[b]Note:[/b] Only in the editor, changing this property emits the [signal Resource.changed] signal`.

paddy-exe replied:
Done as suggested 👍🏻 The signal is only emitted in the editor for performance reasons. This was discussed during rendering meeting with Clay and QbieShay

Geometror reviewed (commented):
Tested the PR locally, everything works except changing the emission shape offset/scale which emits the wrong signal (see comment below).

Small nitpick since this is undefined behavior; when setting the emission ring radius to 0 and the emission ring inner radius to something > 0, one of the rings disappears:

![image](https://github.com/user-attachments/assets/83b9ffed-66ce-4e38-b738-f8d88ceffef6)

I mean we shouldn't allow this at all (add checks to the setters of ParticleProcessMaterial accordingly), but that's for another PR.

AThousandShips commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,323 @@
+				for (int i = 0; i <= 120; i++) {
+					float ra = Math::deg_to_rad((float)(i * 3));
+					float rb = Math::deg_to_rad((float)((i + 1) * 3));
+					Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
```
`sin(ra)` etc. can be pre-computed here when reused so much

AThousandShips commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,323 @@
+			for (int i = 0; i <= 120; i++) {
+				float ra = Math::deg_to_rad((float)(i * 3));
+				float rb = Math::deg_to_rad((float)((i + 1) * 3));
+				Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
```
Same here

AThousandShips commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,323 @@
+
+			for (int i = 0; i <= 120; i = i + 30) {
+				float ra = Math::deg_to_rad((float)(i * 3));
+				Point2 a = Vector2(Math::sin(ra), Math::cos(ra)) * ring_radius;
```
Same here

paddy-exe commented on scene/resources/particle_process_material.cpp:
```diff
@@ -1567,16 +1567,25 @@ void ParticleProcessMaterial::set_emission_shape(EmissionShape p_shape) {
 	emission_shape = p_shape;
 	notify_property_list_changed();
 	_queue_shader_change();
+#ifdef TOOLS_ENABLED
+	emit_changed();
+#endif
```
In that case cc @clayjohn @QbieShay

Mickeon replied:
Indeed, there's a conflict here because it was agreed in a meeting to remove the separate signal and use `changed` instead.

paddy-exe replied:
Okay I changed it back to the `emission_shape_changed` signal with still editor flags around it. I don't see a reason to have it for runtime as well since the visualization is editor-only

AThousandShips commented on doc/classes/ParticleProcessMaterial.xml:
```diff
@@ -187,6 +187,7 @@
 		<member name="emission_box_extents" type="Vector3" setter="set_emission_box_extents" getter="get_emission_box_extents">
 			The box's extents if [member emission_shape] is set to [constant EMISSION_SHAPE_BOX].
 			[b]Note:[/b] [member emission_box_extents] starts from the center point and applies the X, Y, and Z values in both directions. The size is twice the area of the extents.
+			[b]Note:[/b] Only in the editor, changing this property emits the [signal Resource.changed] signal.
```
```suggestion
			[b]Note:[/b] When running in the editor, changing this property emits the [signal Resource.changed] signal.
```
Or similar, to me this wording can be read as if this property is editor exclusive

paddy-exe replied:
Sounds good to me

AThousandShips reviewed (approved):
Looks great! Haven't tested but code looks good

paddy-exe commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,323 @@
+					points.push_back(Vector3(b.x * scale.x + offset.x, b.y * scale.y + offset.y, offset.z));
+				}
+
+				if (p_gizmo->is_selected()) {
+					p_gizmo->add_lines(points, material);
+				}
```
This will differ once we add handles later on. Sometimes we have this here:
```cpp
if (p_gizmo->is_selected()) {
    p_gizmo->add_lines(points, material);
    p_gizmo->add_handles(handles, handles_material, ids);
}
```
 and other times this here:
```cpp
if (p_gizmo->is_selected()) {
    p_gizmo->add_lines(lines, material);	
    p_gizmo->add_handles(handles, handles_material);
}
 ```
 
 See the original PR [here](https://github.com/godotengine/godot/pull/86902/files)
 
 Should I still make the changes? It will be edited later on anyways though

Geometror replied:
I guess it's fine then.

paddy-exe commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,323 @@
+void Particles3DEmissionShapeGizmoPlugin::commit_handle(const EditorNode3DGizmo *p_gizmo, int p_id, bool p_secondary, const Variant &p_restore, bool p_cancel) {
+}
+
+void Particles3DEmissionShapeGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
```
Should be done now I think

Geometror replied:
There are still many more left inside the loops :)
E.g. lines 163-176 can all be `const`

paddy-exe replied:
Done :) thanks for the feedback again!

Geometror commented on scene/resources/particle_process_material.cpp:
```diff
@@ -1684,6 +1708,7 @@ real_t ParticleProcessMaterial::get_emission_ring_cone_angle() const {
 void ParticleProcessMaterial::set_emission_shape_offset(const Vector3 &p_emission_shape_offset) {
 	emission_shape_offset = p_emission_shape_offset;
 	RenderingServer::get_singleton()->material_set_param(_get_material(), shader_names->emission_shape_offset, p_emission_shape_offset);
+	emit_signal("emission_shape_changed");
```
```suggestion
#ifdef TOOLS_ENABLED
	emit_signal("emission_shape_changed");
#endif
```

Geometror commented on scene/resources/particle_process_material.cpp:
```diff
@@ -1693,6 +1718,7 @@ Vector3 ParticleProcessMaterial::get_emission_shape_offset() const {
 void ParticleProcessMaterial::set_emission_shape_scale(const Vector3 &p_emission_shape_scale) {
 	emission_shape_scale = p_emission_shape_scale;
 	RenderingServer::get_singleton()->material_set_param(_get_material(), shader_names->emission_shape_scale, p_emission_shape_scale);
+	emit_signal("emission_shape_changed");
```
```suggestion
#ifdef TOOLS_ENABLED
	emit_signal("emission_shape_changed");
#endif
```

Geometror commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,339 @@
+				p_gizmo->add_lines(points, material);
+			}
+		} else if (shape == CPUParticles3D::EMISSION_SHAPE_BOX) {
+			Vector3 box_extents = particles->get_emission_box_extents();
```
```suggestion
			const Vector3 box_extents = particles->get_emission_box_extents();
```

Geometror commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,339 @@
+				p_gizmo->add_lines(lines, material);
+			}
+		} else if (shape == CPUParticles3D::EMISSION_SHAPE_RING) {
+			float ring_height = particles->get_emission_ring_height();
+			float half_ring_height = ring_height / 2;
+			float ring_radius = particles->get_emission_ring_radius();
+			float ring_inner_radius = particles->get_emission_ring_inner_radius();
+			Vector3 ring_axis = particles->get_emission_ring_axis();
+			float ring_cone_angle = particles->get_emission_ring_cone_angle();
+			float ring_radius_top = MAX(ring_radius - Math::tan(Math::deg_to_rad(90.0 - ring_cone_angle)) * ring_height, 0.0);
+			float ring_inner_radius_top = (ring_inner_radius / ring_radius) * ring_radius_top;
```
```suggestion
			const float ring_height = particles->get_emission_ring_height();
			const float half_ring_height = ring_height / 2;
			const float ring_radius = particles->get_emission_ring_radius();
			const float ring_inner_radius = particles->get_emission_ring_inner_radius();
			const Vector3 ring_axis = particles->get_emission_ring_axis();
			const float ring_cone_angle = particles->get_emission_ring_cone_angle();
			const float ring_radius_top = MAX(ring_radius - Math::tan(Math::deg_to_rad(90.0 - ring_cone_angle)) * ring_height, 0.0);
			const float ring_inner_radius_top = (ring_inner_radius / ring_radius) * ring_radius_top;
```

Geometror commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,339 @@
+			Vector3 box_extents = particles->get_emission_box_extents();
+			Ref<BoxMesh> box;
+			box.instantiate();
+			AABB box_aabb = box->get_aabb();
```
```suggestion
			const AABB box_aabb = box->get_aabb();
```

Geometror commented on editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,339 @@
+				const Vector3 box_extents = mat->get_emission_box_extents();
+				Ref<BoxMesh> box;
+				box.instantiate();
+				AABB box_aabb = box->get_aabb();
```
```suggestion
				const AABB box_aabb = box->get_aabb();
```

Geometror reviewed (approved):
Retested to ensure nothing broke. Great work!
Just a few missing `const`s left, then I think it's good to go.

paddy-exe commented:
Done now and thanks for the detailed feedback 👍🏻

Calinou reviewed (approved):
Tested locally, it works as expected with both GPUParticles3D and CPUParticles3D.

Code and docs look good to me.

While testing, I found an unrelated bug: https://github.com/godotengine/godot/issues/100946

akien-mga commented:
Thanks!

akien-mga commented on scene/3d/gpu_particles_3d.cpp:
```diff
@@ -127,6 +127,7 @@ void GPUParticles3D::set_process_material(const Ref<Material> &p_material) {
 	RID material_rid;
 	if (process_material.is_valid()) {
 		material_rid = process_material->get_rid();
+		process_material->connect("emission_shape_changed", callable_mp((Node3D *)this, &GPUParticles3D::update_gizmos));
```
This doesn't check that `process_material` is a `ParticlesProcessMaterial`, so it fails when it's a `ShaderMaterial`.

paddy-exe replied:
@akien-mga Fix PR is up: https://github.com/godotengine/godot/pull/101617
